rfcs/rfc-005-branches-tags-and-channels.md

```diff
@@ -0,0 +1,48 @@
+
+## Branches and Tags
+
+```
+============================ master branch ============================>
+         \                      \                     \
+          \___v0.7.0 tag         \         v0.9.0 tag__\
+           \                      \                     \_v0.9.1 tag
+            \          v0.8.0 tag__\                     \
+ v0.7.1 tag__\                      \                     \_v0.9.2 tag
+              \___v0.7.2 tag         \___v0.8.1 tag        \
+               \                      \                     \
+                \                      \                     \
+           v0.7 branch             v0.8 branch          v0.9 branch
+
+```
+
+### master branch
+All new development occurs on the `master` branch.
+
+Bug fixes that affect a `vX.Y` branch are also first made on `master`.
+Once the bug fix lands on `master` it is cherry-picked into the `vX.Y` branch.
+The exception to this rule is when a bug fix doesn't apply to `master`.
+
+### v*X.Y* milestone branches
+These are stabilization branches for the given milestone.  They are created off
+the `master` branch as late as possible prior to the milestone release.
+
+It is expected that only the newest milestone branch will be developed on.
+Previous milestone branches remain as a historical curiosity for future code
+archaeologists.
+
+### v*X.Y.0* release tag
+The .0 release tags is created once the milestone is ready for release.
+
+### v*X.Y.1*, v*X.Y.2*, ..., v*X.Y.n* release tags
+Subsequent release tags on a milestone branch are reserved for critical bug
+fixes only.
+
+## Channels
+
+Release channels are used by end-users (humans and bots) to consume the
+branches and tags described in the previous section.
+
+There are three release channels that map into the branches and tags as follows:
+* edge - tracks the `master` branch
+* beta - tracks the largest `vX.Y` branch
+* stable - tracks the largest `vX.Y.Z` tag
```

Comment:
I'd prefer to have "stable" something that's been parked in "beta" for at least 3 weeks.  In such a scheme, stable would track the largest {beta.X}-1 branch, the tags serve only to indicate we've made patches on the branch.

Reply:
Let’s take v0.7 as an example. The stable channel is using the v0.7.1 tag, beta channel is using the v0.7 branch (pretending for a moment it exists).  We wouldn’t cut a v0.7.2 tag, updating the stable channel, until all changes on the v0.7 branch since v0.7.2 was tagged have had that 3 week soak time on the beta channel.   Does that work?

Reply:
six of one, half dozen of the other.  3 weeks from now, we'll have a beta off 0.8.  2 weeks (or so?) later, we'll have 0.9.  at that point, I'd like to say that 0.8 is our "stable".

keeping channels on their own branches also allows us to fix up beta without touching or affecting where stable's gonna be.  stable will be beta after beta's had it's stabilization period.

Reply:
> 2 weeks (or so?) later, we'll have 0.9. at that point, I'd like to say that 0.8 is our "stable".

We'd have the branches v0.8 and v0.9, with some tags v0.8.0 and say v0.8.1 too.
Beta channel is then v0.9, stable is v0.8.1.   So I think you can say 0.8 is our "stable".  If we absolutely *must* make a v0.8.2 at this time (unlikely), we still can since the v0.8 branch still exists.

Are you thinking about a pure train branch model?
```
master ---------------------->
         \      \     \
beta      ----   ----  ----
              \      \     \
release        -----. ----. ---.
```
I'm actively trying to avoid this actually, because it requires that you rewrite history on the beta and release branches as a part of branch promotion.  

> the plan as proposed does not accommodate beta and stable having their own histories 

I don't follow this point.  Can you explain more please?

Reply:
not pure train, at least I don't think so...

let's go back to the original text of the document, which I'd change to:

```
* edge - tracks the `master` branch
* beta - tracks the next-to-largest `vX.Y` branch
* stable - tracks the largest `vX.Y` branch
```

Chromium ups the release number on master when they pick the beta branch.  Under this model we'd be calling master 0.8.0.  On 9/14, master becomes 0.9.0.  "Beta" CI and product pipelines get pointed to a branch called 0.8.0 (or 0.8.1), and "Stable" pipelines get pointed to 0.7.1 (or whatever patchlevel the branch 0.7.X is at).

Cherry picking down isn't a thing we'd do.

Reply:
You have beta/stable swapped right?  ITYM:
```
* edge - tracks the `master` branch
* beta - tracks the largest `vX.Y` branch
* stable - tracks the next-to-largest `vX.Y` branch
```

That seems fine to me.

> Chromium ups the release number on master when they pick the beta branch

Yes. We don't really do that right now but we need to.  I didn't document that yet, figured it was implicit.   But I'll add a section about version bumping to this RFC:
* Incrementing the major version of the master branch is outside the scope of this document.
* Immediately *after* a new beta branch is forged, the minor version of the master branch is incremented.
* Immediately *after* a new stable branch tag has forged, the patch version of the stable branch is incremented.